hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/HiveHoodieReaderContext.java

```diff
@@ -265,10 +266,10 @@ public UnaryOperator<ArrayWritable> projectRecord(Schema from, Schema to, Map<St
   @Override
   public Comparable castValue(Comparable value, Schema.Type newType) {
     //TODO: [HUDI-8261] actually do casting here
-    if (newType == Schema.Type.STRING) {
-      return value.toString();
+    if (value instanceof WritableComparable) {
+      return value;
     }
-    return value;
+    return (WritableComparable) HoodieRealtimeRecordReaderUtils.avroToArrayWritable(value, Schema.create(newType));
```

Comment:
Is this all we need for fixing HUDI-8261 too?  If so, could we make HUDI-8261 fix version 1.0.0 and mark it as closed after this PR is merged?  And remove `//TODO: [HUDI-8261] actually do casting here`.

Reply:
No, it doesn't fix everything, but I think it might be worth it to fix that here as well

Reply:
Spent some time trying to fix 8261. I think it's too complex to get into 1.0.0